crates/lsp/src/header_reader.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    io, mem,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{AsyncBufRead, Future};
+
+const HEADER_DELIMITER: &'static [u8; 4] = b"\r\n\r\n";
+
+pub fn read_headers<'a, R>(
+    reader: &'a mut R,
+    buf: &'a mut Vec<u8>,
+) -> impl Future<Output = io::Result<usize>> + 'a
+where
+    R: AsyncBufRead + ?Sized + Unpin,
+{
+    HeaderReader {
+        reader,
+        buf,
+        read: 0,
+    }
+}
+
+struct HeaderReader<'a, R: ?Sized> {
+    reader: &'a mut R,
+    buf: &'a mut Vec<u8>,
+    read: usize,
+}
+
+impl<'a, R: ?Sized> HeaderReader<'a, R> {
+    fn read_headers<C: AsyncBufRead + ?Sized>(
+        mut reader: Pin<&mut C>,
+        cx: &mut Context<'_>,
+        buf: &mut Vec<u8>,
+        read: &mut usize,
+    ) -> Poll<io::Result<usize>> {
+        loop {
+            let (done, used) = {
+                let available = futures::ready!(reader.as_mut().poll_fill_buf(cx))?;
+                if let Some(i) = find_delimiter(&available) {
+                    buf.extend_from_slice(&available[..i + 5]);
+                    (true, i + 5)
+                } else if buf.len() > 2
+                    && available.len() > 0
+                    && buf[buf.len() - 3..] == HEADER_DELIMITER[..3]
+                    && available[0] == HEADER_DELIMITER[3]
+                {
+                    buf.extend_from_slice(&available[..1]);
+                    (true, 1)
+                } else if buf.len() > 1
+                    && available.len() > 0
+                    && buf[buf.len() - 2..] == HEADER_DELIMITER[..2]
+                    && available[..2] == HEADER_DELIMITER[..2]
+                {
+                    buf.extend_from_slice(&available[..2]);
+                    (true, 2)
+                } else if buf.len() > 0
+                    && available.len() > 0
+                    && buf[buf.len() - 1..] == HEADER_DELIMITER[..1]
+                    && available[..3] == HEADER_DELIMITER[1..3]
+                {
+                    buf.extend_from_slice(&available[..3]);
+                    (true, 3)
+                } else {
+                    buf.extend_from_slice(available);
+                    (false, available.len())
+                }
+            };
```

Comment:
I think this logic could written more clearly with a small `async fn`, instead of a manual implementation of `Future`:

```rust
/// Read from stdout
async fn read_headers(reader: &mut BufReader<Stdout>, buffer: &mut Vec<u8>) -> bool {
    loop {
        if buffer.len() >= HEADER_DELIMITER.len()
            && buffer[(buffer.len() - HEADER_DELIMITER.len())..] == HEADER_DELIMITER
        {
            return true;
        }
        if stdout.read_until(b'\n', &mut buffer).await? == 0 {
            return false;
        }
    }
}
```

Reply:
Oh wow, if I was overthinking it! Thank you for putting me back on the right path, @maxbrunsfeld 🙏